backend/dal.keycloak/Partials/PimsKeycloakUserService.cs

```diff
@@ -117,56 +115,80 @@ public async Task<Entity.User> GetUserAsync(Guid id)
         /// <returns></returns>
         public async Task<Entity.User> UpdateUserAsync(Entity.User user)
         {
-            var kuser = await _keycloakService.GetUserAsync(user.Id) ?? throw new KeyNotFoundException("User does not exist in Keycloak");
-            var euser = _pimsAdminService.User.Get(user.Id);
+            var kuser = await _keycloakService.GetUserAsync(user.Key) ?? throw new KeyNotFoundException("User does not exist in Keycloak");
+            var euser = _pimsAdminService.User.Get(user.Key);
 
             if (user.Username != kuser.Username) throw new InvalidOperationException($"Cannot change the username from '{kuser.Username}' to '{user.Username}'.");
 
-            var addRoles = user.Roles.Except(euser.Roles, new UserRoleRoleIdComparer()).ToArray();
-            var removeRoles = euser.Roles.Except(user.Roles, new UserRoleRoleIdComparer()).ToArray();
-            var addAgencies = user.Agencies.Except(euser.Agencies, new UserAgencyAgencyIdComparer()).ToArray();
-            var removeAgencies = euser.Agencies.Except(user.Agencies, new UserAgencyAgencyIdComparer()).ToArray();
+            IEnumerable<long> addRoleIds;
+            IEnumerable<long> removeRoleIds;
+            if (user.Roles.Any() || euser.Roles.Any())
```

Comment:
So if a user or entity user has no roles they may still have RolesManyToMany populated?

Reply:
This is something we'll need to clean up later.  In the DB, no.  However depending on what we `.Include(...)` in our EF queries, yes.

I've had to keep the ManyToMany reference in the entity for the following reasons.

- MOTI wants to include specific columns for auditing on the many-to-many tables
- EF won't let me not include the property/collection on the parent objects
- All of our code was originally created with the ManyToMany relationships, which means it's easier to continue using it in most places

I would like to make a general rule moving forward that we don't use the ManyToMany tables unless we have to in the future.  Slowly dealing with technical debt as we go.